### PR TITLE
workspace, mappings and fyle changes for V2

### PR DESCRIPTION
**Pull request by Sravanksk**

### CODE REVIEW

File: apps/fyle/models.py [hunk 1]
```diff
@@ -104,38 +106,105 @@ class Meta:
...
-    def create_expense_groups_by_report_id(expense_objects: List[Expense], workspace_id):
+    def create_expense_groups_by_report_id_fund_source(expense_objects: List[Expense], workspace_id):
         """
-        Group expense by report_id
+        Group expense by report_id and fund_source
         """
-        expense_groups = groupby(
-            expense_objects, lambda expense: (expense.report_id, expense.employee_email, expense.claim_number)
-        )
+        department_setting: MappingSetting = MappingSetting.objects.filter(
```
Comment (Shwetabhk) [inline apps/fyle/models.py:113]: why is this needed?
Makes sense for qbo, but why here?

File: apps/fyle/models.py [hunk 2]
```diff
@@ -104,38 +106,105 @@ class Meta:
...
-            report_id = expense_group[0]
-            employee_email = expense_group[1]
-            claim_number = expense_group[2]
+        reimbursable_expenses = list(filter(lambda expense: expense.fund_source == 'PERSONAL', expense_objects))
 
-            expense_ids = Expense.objects.filter(report_id=report_id).values_list(
-                'id', flat=True
+        ccc_expenses = list(filter(lambda expense: expense.fund_source == 'CCC', expense_objects))
+
+        if department_setting and general_settings.reimbursable_expenses_object != 'JOURNAL_ENTRY':
```
Comment (Shwetabhk) [inline apps/fyle/models.py:124]: why are we grouping by Department?

File: apps/fyle/models.py [hunk 3]
```diff
@@ -104,38 +106,105 @@ class Meta:
...
-                fyle_group_id=report_id,
-                workspace_id=workspace_id,
-                defaults={
-                    'description': {
-                        'employee_email': employee_email,
-                        'claim_number': claim_number
+        group_types = [reimbursable_expense_groups]
+
+        if general_settings.corporate_credit_card_expenses_object and ccc_expenses:
+            if department_setting and general_settings.corporate_credit_card_expenses_object != 'JOURNAL_ENTRY':
```
Comment (Shwetabhk) [inline apps/fyle/models.py:143]: why are we grouping by department?

Comment (Sravanksk) [reply apps/fyle/models.py]: All the above comments are resolved now.

File: apps/workspaces/tasks.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+from apps.workspaces.models import WorkspaceSchedule, FyleCredential, WorkspaceGeneralSettings
+
+
+def schedule_sync(workspace_id: int, schedule_enabled: bool, hours: int, next_run: str, user: str):
+    ws_settings, _ = WorkspaceSchedule.objects.get_or_create(
+        workspace_id=workspace_id
+    )
+    start_datetime = datetime.strptime(next_run, '%Y-%m-%dT%H:%M:%S.%fZ')
+
+    if not schedule_enabled:
```
Comment (Shwetabhk) [inline apps/workspaces/tasks.py:19]: Isn't this condition wrong?
This will create a schedule when schedule_enabled is false

File: apps/workspaces/tasks.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+from apps.fyle.models import ExpenseGroup
+from apps.fyle.tasks import create_expense_groups
+from apps.fyle.utils import FyleConnector
+from apps.netsuite.tasks import schedule_bills_creation, schedule_journal_entry_creation, schedule_expense_reports_creation
+from apps.tasks.models import TaskLog
+from apps.workspaces.models import WorkspaceSchedule, FyleCredential, WorkspaceGeneralSettings
+
+
+def schedule_sync(workspace_id: int, schedule_enabled: bool, hours: int, next_run: str, user: str):
```
Comment (Shwetabhk) [inline apps/workspaces/tasks.py:13]: This function needs a lot of refactoring, this was complex when we were using Workspace Settings now this needs to be reimplemented

File: apps/workspaces/views.py [hunk 6]
```diff
@@ -315,3 +320,106 @@ def get(self, request, **kwargs):
...
+        """
+        Scheduled sync
+        """
+        run_sync_schedule(kwargs['workspace_id'], request.user)
+        return Response(
+            status=status.HTTP_200_OK
+        )
+
+
+class SettingsView(viewsets.ViewSet):
```
Comment (Shwetabhk) [inline apps/workspaces/views.py:339]: This is no correct, now since we only have Schedule we should only have schedule view

File: apps/workspaces/views.py [hunk 7]
```diff
@@ -315,3 +320,106 @@ def get(self, request, **kwargs):
...
+    """
+    Settings View
+    """
+    def post(self, request, **kwargs):
```
Comment (Shwetabhk) [inline apps/workspaces/views.py:343]: the whole creation and deletion of jobs / schedule needs to be refactored

File: apps/netsuite/models.py [hunk 8]
```diff
@@ -192,3 +193,284 @@ def create_bill_lineitems(expense_group: ExpenseGroup):
...
+
+        expense = expense_group.expenses.first()
+
+        general_mappings = GeneralMapping.objects.get(workspace_id=expense_group.workspace_id)
+        subsidiary_mappings = SubsidiaryMapping.objects.get(workspace_id=expense_group.workspace_id)
+
+        debit_account_id = None
+
+        entity = Mapping.objects.get(
+            Q(destination_type='EMPLOYEE'),
```
[the remark this answers is not in the document]
Comment (Sravanksk) [reply apps/netsuite/models.py:236]: This works

File: apps/netsuite/utils.py [hunk 9]
```diff
@@ -160,6 +181,26 @@ def sync_vendors(self):
...
 
+    def sync_employees(self):
+        """
+        Sync employees
+        """
+        employees = self.connection.employees.get_all()
+
+        employee_attributes = []
+
+        for employee in employees:
```
[the remark this answers is not in the document]
Comment (Sravanksk) [reply apps/netsuite/utils.py:192]: Will add subsidiary check here also

File: apps/workspaces/views.py (hunk 6 above)
Comment (Sravanksk) [reply apps/workspaces/views.py:339]: Will rename this to `ScheduleView`

File: apps/workspaces/views.py (hunk 7 above)
Comment (Sravanksk) [reply apps/workspaces/views.py:343]: Will update the whole `schedule` part

File: apps/mappings/models.py [hunk 10]
```diff
@@ -29,12 +29,16 @@ class GeneralMapping(models.Model):
...
+    reimbursable_account_name = models.CharField(max_length=255, help_text='Reimbursable Expenses Account name', null=True)
+    reimbursable_account_id = models.CharField(max_length=255, help_text='Reimbursable Expenses Account id', null=True)
+    default_ccc_account_name = models.CharField(max_length=255, help_text='CCC Expenses Account name', null=True)
+    default_ccc_account_id = models.CharField(max_length=255, help_text='CCC Expenses Account id', null=True)
     workspace = models.ForeignKey(Workspace, on_delete=models.PROTECT, help_text='Reference to Workspace model')
     created_at = models.DateTimeField(auto_now_add=True, help_text='Created at datetime')
     updated_at = models.DateTimeField(auto_now=True, help_text='Updated at datetime')
 
     class Meta:
-        unique_together = ('location_name', 'accounts_payable_name', 'workspace')
```
[the remark this answers is not in the document]
Comment (Sravanksk) [reply apps/mappings/models.py:39]: Don't remember why, but this won't create any issues now.

File: apps/mappings/urls.py [hunk 11]
```diff
@@ -1,18 +1,3 @@
-"""fyle_qbo URL Configuration
-
```
[the remark this answers is not in the document]
Comment (Sravanksk) [reply apps/mappings/urls.py:2]: Will add a description to all URL files

File: apps/netsuite/models.py [hunk 12]
```diff
@@ -192,3 +193,284 @@ def create_bill_lineitems(expense_group: ExpenseGroup):
...
+            defaults={
+                'account_id': debit_account_id,
+                'entity_id': entity.destination.destination_id,
+                'currency': expense.currency,
+                'department_id': None,
+                'class_id': None,
+                'location_id': general_mappings.location_id,
+                'subsidiary_id': subsidiary_mappings.internal_id,
+                'memo': 'Report {0} / {1} exported on {2} for {3} '.format(
+                    expense.claim_number, expense.report_id, datetime.now().strftime("%Y-%m-%d"), expense.purpose
```
[the remark this answers is not in the document]
Comment (Sravanksk) [reply apps/netsuite/models.py:264]: Yes
```
'memo': 'Report {0} / {1} exported on {2}'.format(expense.claim_number, expense.report_id, datetime.now().strftime("%Y-%m-%d")
```

`Report C/2020/05/R/5 / rpHetNXsfMmR exported on 2020-07-23`

Does this memo work?

File: apps/netsuite/utils.py [hunk 13]
```diff
@@ -43,6 +43,9 @@ def sync_accounts(self, account_type):
         elif account_type == '_accountsPayable':
             attribute_type = 'ACCOUNTS_PAYABLE'
             display_name = 'Accounts Payable'
+        elif account_type == '_creditCard':
+            attribute_type = 'CREDIT_CARD_ACCOUNT'
+            display_name = 'Credit Card Account'
         else:
             attribute_type = None
```
[the remark this answers is not in the document]
Comment (Sravanksk) [reply apps/netsuite/utils.py:50]: Will refactor sync_accounts since it's taking lots of time to fetch accounts

File: apps/netsuite/views.py [hunk 14]
```diff
@@ -50,6 +58,17 @@ def post(self, request, *args, **kwargs):
...
                 status=status.HTTP_400_BAD_REQUEST
             )
+        except NetSuiteRequestError as exception:
+            logger.exception(exception)
+            detail = json.dumps(exception.__dict__)
+            detail = json.loads(detail)
+
+            return Response(
+                data={
+                    'message': detail['code'] + detail['message']
```
Comment (Shwetabhk) [inline apps/netsuite/views.py:68]: use string formatting everywhere

File: apps/netsuite/views.py [hunk 15]
```diff
@@ -85,20 +104,47 @@ def post(self, request, *args, **kwargs):
...
                 status=status.HTTP_400_BAD_REQUEST
             )
+        except NetSuiteRequestError as exception:
+            logger.exception(exception)
+            detail = json.dumps(exception.__dict__)
+            detail = json.loads(detail)
 
+            return Response(
+                data={
+                    'message': detail['code'] + detail['message']
```
Comment (Shwetabhk) [inline apps/netsuite/views.py:114]: string formatting

File: apps/netsuite/views.py [hunk 16]
```diff
@@ -85,20 +104,47 @@ def post(self, request, *args, **kwargs):
...
+                },
+                status=status.HTTP_401_UNAUTHORIZED
+            )
 
-class AccountView(generics.ListCreateAPIView):
+
+class AccountView(viewsets.ViewSet):
```
Comment (Shwetabhk) [inline apps/netsuite/views.py:120]: lets not use viewset here
Keep it as separate views

File: apps/netsuite/models.py [hunk 17]
```diff
@@ -192,3 +192,284 @@ def create_bill_lineitems(expense_group: ExpenseGroup):
...
+        """
+        expenses = expense_group.expenses.all()
+        journal_entry = JournalEntry.objects.get(expense_group=expense_group)
+
+        description = expense_group.description
+
+        debit_account_id = None
+
+        entity = Mapping.objects.get(
+            destination_type='EMPLOYEE' or 'VENDOR',
```
Comment (Shwetabhk) [inline apps/netsuite/models.py:418]: This is wrong
should be using `Q` here

Comment (Shwetabhk) [reply apps/netsuite/models.py:418]: This is not how `or` works in Django ORM
